book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,218 @@
+slots.
+
+### Voting Before Catching Up
+
+Any vote that a validator signs must have its lockout observed. If a validator
+has to reboot, the most secure way for it to check if its lockouts apply to the
+new set of forks is to see the votes on the ledger. If it can do this, it knows
+that those votes do not lock it out of voting on the current fork, so it just
+needs to wait for the lockouts of the votes not on the ledger to expire.
+However, if the validator votes before catching up, the votes will not go onto
+the ledger, so if the validator reboots, it will have to assume that the votes
+lock it out from voting again.
+
+## Snapshot Verification Overview
+
+A booting validator needs to have some set of trusted validators whose votes it
+will rely upon to ensure that it has caught up to the cluster and has a valid
+snapshot. With that set, the validator can follow the procedure detailed in the
+`Boot Procedure` section.
+
+For the sections below:
+
+Assume an arbitrary locktower state `L`, a snapshot root `S`, and a trusted
+bank slot `T` (defined in step 7-1 of the `Boot Procedure` section).
+
+Define `L_i` to be: The slot of the ith vote in `L`.
+
+Define the function `is_ancestor(a, b)` to be: Given a slot `a` and a slot `b`
+will return true if `a` is an ancestor of `b`.
+
+Define the function `is_locked(a, b)` to be: `a + a.lockout <= b`
+
+## Determining Vote Safety From a Snapshot and Locktower
+
+Define the "safety" condition to be: 
+
+Given a `BankForks`, a validator can run some procedure to determine whether it
+can vote on a frozen bank `T` (frozen banks must be present in
+`BankForks.banks`), without violating the lockouts of any vote in `L`. The
+procedure for this is:
+
+1) If `T` < last locktower vote, return false,
+
+2) Apply `T` to locktower state `L`. Pop off all votes `S_i` where
+`!is_locked(S_i, T)`. `T` must now be the top of the tower.
+
+3) If for any remaining vote `L_i` in `L` `!is_ancestor(L_i, T)`, return false.
+
+4) Otherwise, return true.
+
+## Achieving Safety
+
+Define the "Safety Criteria" to be: Given a `BankForks` any descendant of `S`,
+`S_d` that has a bank `B_d` that is present in `BankForks.banks`, and any slot
+`L_i` in locktower, the validator is able to determine `is_ancestor(L_i, S_d)`.
+
+Assume the "Safety Criteria" is true, we show we can then achieve the "safety"
+condition:
+
+**Proof:** A validator wants to determine whether it can vote on `T`. `T` must
+be a descendant of `S` because the validator does not play any state for
+non-descendants of `S` when booting from the snapshot. From assuming "Safety
+Criteria" above, this means for any `L_i` we can determine `is_ancestor(L_i,
+T)`. This means we have all the tools to run the algorithm from the safety
+definition.
+
+Thus to achieve safety, we want to design the snapshotting system such that the
+"Safety Criteria" is met.
+
+## Implementing "Safety Criteria"
+
+The snapshot is augmented to store the last `N` ancestors of `S`. These
```

Comment:
Sysvar?

Reply:
N should be 1 epoch I think.

Reply:
I'm not sure what you're asking here. The idea is to store the `N` most recent ancestors of the snapshot root. In the validator this could be a ring buffer storing the ancestors of the most recent root, which then gets serialized when the snapshot is created. It doesn't seem like this needs to be a sysvar.

Reply:
Wow, GitHub only showed me one of @aeyakovenko's two comments... Anyway, `N = DEFAULT_SLOTS_PER_EPOCH` seems reasonable.

Reply:
Any information that is not signed and voted on as part of consensus it can’t be trusted.

Reply:
The next sentence is "These ancestors are incorporated into the bank hash so they can be verified when a validator unpacks a snapshot." Is that not adequate?